### PR TITLE
Add types to qos.py

### PR DESCRIPTION
**Pull request by InvincibleRMC**

### CODE REVIEW

**Review by fujitatomoya — APPROVED** [review]
lgtm with minor comment. i would like to ask 2nd review.

**Comment by fujitatomoya** [comment]
i think we can ignore https://build.ros2.org/job/Rpr__rclpy__ubuntu_noble_amd64/30/testReport/ rpr job that are unrelated.

File: rclpy/rclpy/duration.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 from rclpy.impl.implementation_singleton import rclpy_implementation as _rclpy
 
 
-class DurationType(Protocol):
+class DurationHandle(Protocol):
```
[the remark this answers is not in the document]
Comment (InvincibleRMC) [reply rclpy/rclpy/duration.py:22]: In a past pull request someone had mentioned `DurationType` might be confused with `Type[Duration]` Maybe a name of `RCLDuration` or `DurationT` would be acceptable alternatives?

File: rclpy/rclpy/qos.py [hunk 2]
```diff
@@ -73,96 +87,100 @@ class QoSProfile:
...
-            QoSProfile.__qos_profile_default_dict['liveliness_lease_duration'])
-        self.avoid_ros_namespace_conventions = kwargs.get(
-            'avoid_ros_namespace_conventions',
-            QoSProfile.__qos_profile_default_dict['avoid_ros_namespace_conventions'])
+        if depth is None:
+            self.depth = QoSProfile.__qos_profile_default_dict['depth']
+        else:
+            self.depth = depth
+
+        self.reliability = reliability or QoSProfile.__qos_profile_default_dict['reliability']
```
Comment (sloretz) [inline rclpy/rclpy/qos.py:115]: I think the `or` logic makes it impossible to use the `SYSTEM_DEFAULT` QoS policy, which IIRC is the one that allows the QoS to be set by middleware specific configuration.

```python3
>>> from enum import IntEnum
>>> class ReliabilityPolicy(IntEnum):
...     SYSTEM_DEFAULT = 0
...     RELIABLE = 1
...     BEST_EFFORT = 2
...     UNKNOWN = 3
...     BEST_AVAILABLE = 4
...
>>> ReliabilityPolicy.SYSTEM_DEFAULT or ReliabilityPolicy.RELIABLE
<ReliabilityPolicy.RELIABLE: 1>
```

I'd recommend instead for all of these:

```python3
self.reliability = QoSProfile.__qos_profile_default_dict['reliability'] if reliability is None else reliability
```

**Review by sloretz — CHANGES_REQUESTED** [review]
One fix, then the rest LGTM!

File: rclpy/rclpy/qos.py (hunk 2 above)
Comment (InvincibleRMC) [reply rclpy/rclpy/qos.py:115]: Will do!

**Review by sloretz — APPROVED** [review]

**Comment by sloretz** [comment]
Pulls: ros2/rclpy#1255
Gist: https://gist.githubusercontent.com/sloretz/f3fb2535ace0225ba8d2f32ed9b567ae/raw/85ef195102947295b8d04b9ecb4c33299aa77af6/ros2.repos
BUILD args:  --packages-up-to rclpy
TEST args:  --packages-select rclpy
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14360
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21613/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15901/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1118/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22331/)

**Comment by sloretz** [comment]
Pulls: ros2/rclpy#1255
Gist: https://gist.githubusercontent.com/sloretz/eb8119e338f75379aa3f5b45f73dd26d/raw/85ef195102947295b8d04b9ecb4c33299aa77af6/ros2.repos
BUILD args:  --packages-up-to rclpy
TEST args:  --packages-select rclpy
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14376
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21631/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15917/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1135/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22347/)

**Review by sloretz — APPROVED** [review]
Thank you for the PR!

**Pull request merged**